File: 2023/10.2.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <map>
#include <tuple>
#include <queue>
#include <utility>
#include <tuple>
#include <unordered_map>
#include <algorithm>

using namespace std;

ofstream outputFile;

void setOutputStream()
{
    outputFile.open("output.txt");
    if (!outputFile.is_open())
    {
        exit(EXIT_FAILURE);
    }

    cout.rdbuf(outputFile.rdbuf());
}

void displayGrid(vector<vector<char>> &grid)
{
    for (auto &row : grid)
    {
        for (auto &ch : row)
        {
            cout << ch;
        }
        cout << endl;
    }
}

vector<vector<char>> parseInput()
{
    ifstream inputFile("input.txt");
    if (!inputFile.is_open())
        exit(EXIT_FAILURE);

    string line;
    vector<vector<char>> pipes;
    while (getline(inputFile, line))
    {
        vector<char> singleLine;
        for (auto &ch : line)
            singleLine.push_back(ch);

        pipes.push_back(singleLine);
    }

    // displayGrid(pipes);

    return pipes;
}

pair<int, int> getSourcePos(vector<vector<char>> &grid)
{
    for (size_t i = 0; i < grid.size(); i++)
    {
        for (size_t j = 0; j < grid[0].size(); j++)
        {
            if (grid[i][j] == 'S')
            {
                return make_pair(i, j);
            }
        }
    }

    return make_pair(-1, -1);
}

void fixSourceChar(vector<vector<char>> &grid, int sx, int sy)
{
    int m = grid.size();
    int n = grid[0].size();

    map<vector<int>, char> positions = {
        {{1, 0, 1, 0}, 'J'},
        {{1, 1, 0, 0}, '-'},
        {{1, 0, 0, 1}, '7'},
        {{0, 1, 0, 1}, 'F'},
        {{0, 1, 1, 0}, 'L'},
        {{0, 0, 1, 1}, '|'},
    };

    int left = 0;
    int right = 0;
    int top = 0;
    int bottom = 0;

    if (sy - 1 >= 0 && (grid[sx][sy - 1] == '-' || grid[sx][sy - 1] == 'L' || grid[sx][sy - 1] == 'F'))
    {
        left = 1;
    }
    if (sy + 1 < n && (grid[sx][sy + 1] == '-' || grid[sx][sy + 1] == 'J' || grid[sx][sy + 1] == '7'))
    {
        right = 1;
    }
    if (sx - 1 >= 0 && (grid[sx - 1][sy] == '|' || grid[sx - 1][sy] == '7' || grid[sx - 1][sy] == 'F'))
    {
        top = 1;
    }
    if (sx + 1 < m && (grid[sx + 1][sy] == '|' || grid[sx + 1][sy] == 'L' || grid[sx + 1][sy] == 'J'))
    {
        bottom = 1;
    }

    vector<int> positionsKey = {left, right, top, bottom};

    // for (const auto &element : positionsKey)
    // {
    //     std::cout << element << " ";
    // }

    grid[sx][sy] = positions[positionsKey];
}

int findLongestLoopLength(vector<vector<char>> &grid, vector<vector<int>> &distanceGrid, int sx, int sy)
{
    int m = grid.size();
    int n = grid[0].size();

    unordered_map<char, vector<pair<int, int>>> directions = {
        {'|', {{-1, 0}, {1, 0}}},
        {'-', {{0, -1}, {0, 1}}},
        {'L', {{0, 1}, {-1, 0}}},
        {'J', {{0, -1}, {-1, 0}}},
        {'7', {{0, -1}, {1, 0}}},
        {'F', {{1, 0}, {0, 1}}},
        {'S', {{0, 1}, {1, 0}, {-1, 0}, {0, -1}}}};

    queue<tuple<int, int, int>> que;

    que.push(make_tuple(sx, sy, 0));
    tuple<int, int, int> front;

    while (!que.empty())
    {
        front = que.front();
        que.pop();
        int x = get<0>(front), y = get<1>(front), steps = get<2>(front);

        if (distanceGrid[x][y] != -1)
        {
            continue;
        }
        distanceGrid[x][y] = steps;

        for (auto &pr : directions[grid[x][y]])
        {
            // cout << pr.first << " " << pr.second << endl;
            int dx = pr.first, dy = pr.second;
            int next_x = x + dx, next_y = y + dy;

            if (0 <= next_x && next_x < m && 0 <= next_y && next_y < n && distanceGrid[next_x][next_y] == -1)
            {
                que.push(make_tuple(next_x, next_y, steps + 1));
            }
        }
    }

    int farthest = 0;
    for (auto &row : distanceGrid)
    {
        auto maxInRow = max_element(row.begin(), row.end());
        farthest = max(farthest, static_cast<int>(*maxInRow));
    }

    return farthest;
}

void cleanPipes(vector<vector<char>> &pipes, vector<vector<int>> &distanceGrid)
{
    for (size_t i = 0; i < distanceGrid.size(); i++)
    {
        for (size_t j = 0; j < distanceGrid[i].size(); j++)
        {
            if (distanceGrid[i][j] == -1)
            {
                pipes[i][j] = '.';
            }
        }
    }
}

int rayCastArea(vector<vector<char>> &pipes)
{
    long int area = 0;

    for (size_t i = 0; i < pipes.size(); i++)
    {
        for (size_t j = 0; j < pipes[i].size(); j++)
        {
            // J, F, L, 7, | count
            int cnt[5] = {0};

            if (pipes[i][j] == '.')
            {
                for (size_t k = j + 1; k < pipes[i].size(); k++)
                {
                    if (pipes[i][k] == 'J')
                    {
                        ++cnt[0];
                    }
                    else if (pipes[i][k] == 'F')
                    {
                        ++cnt[1];
                    }
                    else if (pipes[i][k] == 'L')
                    {
                        ++cnt[2];
                    }
                    else if (pipes[i][k] == '7')
                    {
                        ++cnt[3];
                    }
                    else if (pipes[i][k] == '|')
                    {
                        ++cnt[4];
                    }
                }
                int intersections = cnt[4] + min(cnt[0], cnt[1]) + min(cnt[2], cnt[3]);
                if (intersections % 2)
                {
                    ++area;
                }
            }
        }
    }

    return area;
}

int main()
{
    setOutputStream();
    vector<vector<char>> pipes = parseInput();

    // displayGrid(pipes);

    pair<int, int> sourcePos = getSourcePos(pipes);
    int sx = sourcePos.first, sy = sourcePos.second;

    // cout << sx << " " << sy << endl;

    fixSourceChar(pipes, sx, sy);

    // displayGrid(pipes);

    int longestLoopLength = 0;
    vector<pair<int, int>> loopPositions;

    vector<vector<int>> distanceGrid(pipes.size(), vector<int>(pipes[0].size(), -1));
    int loop_length = findLongestLoopLength(pipes, distanceGrid, sx, sy);

    cout << "Loop length: " << loop_length << endl;

    cleanPipes(pipes, distanceGrid);
    // displayGrid(pipes);

    long int area = rayCastArea(pipes);
    cout << "Area: " << area << endl;

    return 0;
}

File: 2023/10.2.py
import collections
import sys
import time
from pprint import pprint

try:
    sys.stdin = open(sys.path[0] + "/input.txt", "r")
    sys.stdout = open(sys.path[0] + "/output.txt", "w")
except FileNotFoundError:
    pass


def parse_input():
    pipes = []

    for line in sys.stdin:
        line = list(line.strip())
        pipes.append(line)

    return pipes


def find_start_pos(pipes):
    for i in range(len(pipes)):
        for j in range(len(pipes[0])):
            if pipes[i][j] == "S":
                return i, j


def is_movement_possible_from_s(dx, dy, next_char):
    # | is a vertical pipe connecting north and south.
    # - is a horizontal pipe connecting east and west.
    # L is a 90-degree bend connecting north and east.
    # J is a 90-degree bend connecting north and west.
    # 7 is a 90-degree bend connecting south and west.
    # F is a 90-degree bend connecting south and east.
    # . is ground; there is no pipe in this tile.
    # S is the starting position of the animal

    possible_movements = {
        # West
        (0, -1): ("-", "L", "F"),
        # East
        (0, 1): ("-", "J", "7"),
        # North
        (-1, 0): ("|", "7", "F"),
        # South
        (1, 0): ("|", "L", "J"),
    }

    return next_char in possible_movements[(dx, dy)]


def display_loop(longest_loop_set, pipes):
    m, n = len(pipes), len(pipes[0])
    for i in range(m):
        print(*[pipes[i][j] if (i, j) in longest_loop_set else "." for j in range(n)])
    print()


def find_longest_loop(pipes, x, y):
    m, n = len(pipes), len(pipes[0])
    longest_loop = []
    longest_loop_length = 0

    # x, y, path, visited
    start_x, start_y = x, y
    stack = [(x, y, [], set())]
    while stack:
        x, y, curr_path, visited = stack.pop()
        if not 0 <= x < m or not 0 <= y < n or pipes[x][y] not in dir_map:
            continue

        # display_loop(set(curr_path), pipes)
        # time.sleep(0.2)
        # Update longest path
        if x == start_x and y == start_y:
            # display_loop(set(curr_path), pipes)
            if len(curr_path) > longest_loop_length:
                longest_loop_length = len(curr_path)
                longest_loop = curr_path

        if (x, y) in visited:
            continue
        visited.add((x, y))

        for dx, dy in dir_map[pipes[x][y]]:
            next_x = x + dx
            next_y = y + dy

            if 0 <= next_x < m and 0 <= next_y < n:
                if pipes[x][y] == "S" and not is_movement_possible_from_s(
                    dx, dy, pipes[next_x][next_y]
                ):
                    continue
                stack.append((next_x, next_y, curr_path + [(x, y)], visited))

    # print("Loop length:", len(longest_loop))

    return longest_loop, longest_loop_length


def fix_source_char(grid, sx, sy):
    # Grid contains only the longest loop
    m = len(grid)
    n = len(grid[0])

    # Key: Left, Right, Top, Bottom
    # 1 means occupied and 0 means unoccupied
    positions = {
        (1, 0, 1, 0): "J",
        (1, 1, 0, 0): "-",
        (1, 0, 0, 1): "7",
        (0, 1, 0, 1): "F",
        (0, 1, 1, 0): "L",
        (0, 0, 1, 1): "|",
    }

    left = 0
    right = 0
    top = 0
    bottom = 0

    if sy - 1 >= 0 and grid[sx][sy - 1] in ("-", "L", "F"):
        left = 1
    if sy + 1 < n and grid[sx][sy + 1] in ("-", "J", "7"):
        right = 1
    if sx - 1 >= 0 and grid[sx - 1][sy] in ("|", "7", "F"):
        top = 1
    if sx + 1 < m and grid[sx + 1][sy] in ("|", "L", "J"):
        bottom = 1

    grid[sx][sy] = positions[(left, right, top, bottom)]


def ray_cast(grid):
    m = len(grid)
    n = len(grid[0])

    # for row in grid:
    #     print(*row)

    area = 0

    for i in range(m):
        for j in range(n):
            if grid[i][j] == 0:
                freq = {
                    "|": 0,
                    "F": 0,
                    "J": 0,
                    "L": 0,
                    "7": 0,
                }
                for k in range(j + 1, n):
                    if grid[i][k] in freq:
                        freq[grid[i][k]] += 1

                intersections = (
                    freq["|"] + min(freq["F"], freq["J"]) + min(freq["L"], freq["7"])
                )
                if intersections % 2:
                    area += 1

    return area


def get_area_inside_loop(pipes, loop_path):
    m, n = len(pipes), len(pipes[0])
    loop_matrix = [[0] * n for _ in range(m)]
    loop_path_set = set(loop_path)

    for i in range(m):
        for j in range(n):
            if (i, j) in loop_path_set:
                loop_matrix[i][j] = pipes[i][j]

    sx, sy = find_start_pos(loop_matrix)
    fix_source_char(loop_matrix, sx, sy)

    return ray_cast(loop_matrix)


pipes = parse_input()

dir_map = {
    "|": [(-1, 0), (1, 0)],
    "-": [(0, -1), (0, 1)],
    "L": [(0, 1), (-1, 0)],
    "J": [(0, -1), (-1, 0)],
    "7": [(0, -1), (1, 0)],
    "F": [(1, 0), (0, 1)],
    "S": [(0, 1), (1, 0), (-1, 0), (0, -1)],
}

# Locate S
start_x, start_y = find_start_pos(pipes)

# Find the longest loop
longest_loop, longest_loop_length = find_longest_loop(pipes, start_x, start_y)

area = get_area_inside_loop(pipes, longest_loop)
print("Area: ", area)
